File: apps/worker/src/executor.ts
import { existsSync } from "fs";
import { getRoute } from "./routes.js";
import type { ConversionRoute } from "@zenvort/shared";

export type AttemptResult = {
  converter: string;
  success: boolean;
  durationMs: number;
  error?: string;
};

/**
 * Returns true if the error is a timeout error (Converter timed out).
 */
export function isTimeoutError(err: unknown): boolean {
  if (err instanceof Error) {
    return (
      err.message.includes("timed out") ||
      err.message.includes("timed out after")
    );
  }
  return false;
}

/**
 * Executes the fallback conversion chain for a job.
 *
 * @returns { converterUsed, attempts } on success
 * @throws Aggregated error if all converters fail
 */
export async function executeConversion(
  jobId: string,
  inputPath: string,
  outputPath: string,
  inputFormat: string,
  outputFormat: string
): Promise<{ converterUsed: string; attempts: AttemptResult[] }> {
  // ── Security checks ─────────────────────────────────────────────────────────
  const UUID_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
  if (!UUID_REGEX.test(jobId)) {
    throw new Error(`executeConversion: jobId is not a valid UUID: ${jobId}`);
  }

  // Paths are validated by sanitizeAndAssertTmpPath() inside each converter (cross-platform).
  const resolvedInput  = inputPath;
  const resolvedOutput = outputPath;

  // ── Route lookup ───────────────────────────────────────────────────────────
  const route: ConversionRoute | null = getRoute(inputFormat, outputFormat);
  if (!route) {
    throw new Error(
      `No conversion route for ${inputFormat}→${outputFormat}`
    );
  }

  const { converters, description } = route;
  const attempts: AttemptResult[] = [];

  console.log(
    `[${jobId}] Starting ${description} (${converters.length} converter(s))`
  );

  // ── Fallback chain ─────────────────────────────────────────────────────────
  for (let i = 0; i < converters.length; i++) {
    const converterFn = converters[i];
    const converterName = converterFn.name || "anonymous";

    console.log(
      `[${jobId}] Attempt ${i + 1}/${converters.length} — ${converterName}`
    );

    // Clean slate before each attempt
    if (existsSync(resolvedOutput)) {
      try {
        await import("fs").then((fs) => fs.promises.unlink(resolvedOutput));
      } catch {
        // Ignore cleanup errors — may not exist
      }
    }

    const start = performance.now();

    try {
      await converterFn(resolvedInput, resolvedOutput, inputFormat, outputFormat);

      const durationMs = Math.round(performance.now() - start);

      console.log(
        `[${jobId}] ${converterName} succeeded in ${durationMs}ms`
      );

      attempts.push({
        converter: converterName,
        success: true,
        durationMs,
      });

      return { converterUsed: converterName, attempts };
    } catch (err) {
      const durationMs = Math.round(performance.now() - start);
      const isTimeout = isTimeoutError(err);
      const errorMsg = err instanceof Error ? err.message : String(err);

      if (isTimeout) {
        console.log(`[${jobId}] ${converterName} timed out after ${durationMs}ms`);
      } else {
        console.log(`[${jobId}] ${converterName} failed: ${errorMsg}`);
      }

      attempts.push({
        converter: converterName,
        success: false,
        durationMs,
        error: errorMsg,
      });

      // If last converter failed, the aggregated error is thrown below
    }
  }

  // ── All converters failed ─────────────────────────────────────────────────
  const errors = attempts
    .map((a) => `  - ${a.converter}: ${a.error ?? "unknown error"}`)
    .join("\n");

  throw new Error(
    `All converters failed for ${inputFormat}→${outputFormat}:\n${errors}`
  );
}

File: worker/executor.py
import os
import time
import uuid
import logging
from dataclasses import dataclass

from worker.security.path_guard import TMP_DIR, sanitize_and_assert_tmp_path
from worker.utils import _sanitize_error

logger = logging.getLogger("zenvort.worker")


@dataclass
class AttemptResult:
    converter_name: str
    duration_ms: float
    error: str | None


def execute_conversion(
    job_id: str,
    input_path: str,
    output_path: str,
    input_format: str,
    output_format: str,
) -> dict:
    """Dispatch a file conversion through the converter chain for the given format pair.

    Tries each converter in the route list sequentially. If a converter fails,
    the next one is tried as a fallback. Only raises if ALL converters fail.

    Args:
        job_id: UUID of the conversion job (validated before dispatch).
        input_path: Absolute path to the input file (must be within TMP_DIR).
        output_path: Absolute path for the output file (must be within TMP_DIR).
        input_format: Source format (e.g. "pdf").
        output_format: Target format (e.g. "docx").

    Returns:
        dict with keys: "converter_used" (str), "attempts" (list[AttemptResult]).

    Raises:
        ValueError: Invalid job_id, or path traversal detected.
        RuntimeError: No route found, or all converters failed.
    """
    try:
        uuid.UUID(job_id)
    except ValueError:
        raise ValueError(f"Invalid job ID: {job_id}")

    sanitize_and_assert_tmp_path(input_path)
    sanitize_and_assert_tmp_path(output_path)

    from worker.routes import ROUTES
    converters = ROUTES.get(f"{input_format}→{output_format}")
    if not converters:
        raise RuntimeError(f"No conversion route for {input_format}→{output_format}")

    attempts: list[AttemptResult] = []

    for converter_fn in converters:
        # Identify converter: prefer __spec__.name (module-installed), fallback
        # to __module__ (inline script-style).  This handles both the 1-file
        # converter style (functions with __module__) and 4-file style
        # (module-installed, where __spec__ is set).
        spec = getattr(converter_fn, "__spec__", None)
        if spec is not None:
            converter_name = f"{spec.name}.{converter_fn.__name__}"
        else:
            converter_name = f"{converter_fn.__module__}.{converter_fn.__name__}"

        if os.path.exists(output_path):
            os.unlink(output_path)

        start = time.perf_counter()
        error: str | None = None

        try:
            converter_fn(input_path, output_path, input_format, output_format)
        except Exception as exc:  # noqa: BLE001 — intentional broad catch for fallback chain
            error = str(exc)

        duration_ms = int((time.perf_counter() - start) * 1000)

        if error is None:
            return {
                "converter_used": converter_name,
                "attempts": attempts
                + [AttemptResult(converter_name, duration_ms, None)],
            }

        attempts.append(AttemptResult(converter_name, duration_ms, error))
        logger.info(f"[executor] {converter_name} failed: {error}")

    # All converters failed — log internal details for ops, return sanitized for user
    logger.error(
        f"All converters failed for {input_format}→{output_format}:\n"
        + "\n".join(f"  {a.converter_name}: {a.error}" for a in attempts)
    )
    raise RuntimeError(
        _sanitize_error(
            "\n".join(attempt.error or "" for attempt in attempts),
            input_format,
            output_format,
        )
    )